Document Node.js Crypto API support (#10067)
cloudflare/cloudflare-docs: content/workers/runtime-apis/nodejs/crypto.md

---
pcx_content_type: configuration
title: Crypto
---

# Crypto

{{<render file="_nodejs-compat-howto.md">}}

The `node:crypto`` module provides cryptographic functionality that includes a set of wrappers for OpenSSL's hash, HMAC, cipher, decipher, sign, and verify functions.

> A subset of the the `node:crypto` module is available in Workers. All APIs in the tables below with a ✅ are supported, and unless otherwise noted, work the same way as the implementations in Node.js.

Author: ```suggestion
A subset of the the `node:crypto` module is available in Workers. All APIs in the tables below with a ✅ are supported, and unless otherwise noted, work the same way as the implementations in Node.js.

The Workers implementation of `node:crypto` always runs in a FIPS-compliant mode. As such, APIs such as `setFips()` will not be made available.
```

| [KeyObject](https://nodejs.org/api/crypto.html#class-keyobject) |  | |
| [Sign](https://nodejs.org/api/crypto.html#class-sign) |  | |
| [Verify](https://nodejs.org/api/crypto.html#class-verify) |  | |
| [X509Certificate](https://nodejs.org/api/crypto.html#class-x509certificate) |  | |
| [constants](https://nodejs.org/api/crypto.html#cryptoconstants) |  | |

## Primes

| API | Supported? | Notes |
|-----|------------|-------|
| [checkPrime](https://nodejs.org/api/crypto.html#cryptocheckprimecandidate-options-callback) | ✅ | |
| [checkPrimeSync](https://nodejs.org/api/crypto.html#cryptocheckprimesynccandidate-options) | ✅ | |
| [generatePrime](https://nodejs.org/api/crypto.html#cryptogenerateprimesize-options-callback) | ✅ | |
| [generatePrimeSync](https://nodejs.org/api/crypto.html#cryptogenerateprimesyncsize-options) | ✅ | |

## Ciphers

| API | Supported? | Notes |
|-----|------------|-------|
| [createCipher](https://nodejs.org/api/crypto.html#cryptocreatecipheralgorithm-password-options) |  | Deprecated, use `createCipheriv` instead |
| [createCipheriv](https://nodejs.org/api/crypto.html#cryptocreatecipherivalgorithm-key-iv-options) |  | |

> | [createDecipher](https://nodejs.org/api/crypto.html#cryptocreatedecipheralgorithm-password-options) |  | Deprecated, use `createDecipheriv` instead |

Author: ```suggestion
| [createDecipher](https://nodejs.org/api/crypto.html#cryptocreatedecipheralgorithm-password-options) |  ❌  | Deprecated, will not be implemented. Use `createDecipheriv` instead. |
```

| [generateKeySync](https://nodejs.org/api/crypto.html#cryptogeneratekeysynctype-options) | ✅ | |

## Sign/Verify

| API | Supported? | Notes |
|-----|------------|-------|
| [createSign](https://nodejs.org/api/crypto.html#cryptocreatesignalgorithm-options) |  | |
| [createVerify](https://nodejs.org/api/crypto.html#cryptocreateverifyalgorithm-options) |  | |
| [sign](https://nodejs.org/api/crypto.html#cryptosignalgorithm-data-key-callback) |  | |
| [verify](https://nodejs.org/api/crypto.html#cryptoverifyalgorithm-data-key-signature-callback) |  | |

## Misc

| API | Supported? | Notes |
|-----|------------|-------|
| [getCipherInfo](https://nodejs.org/api/crypto.html#cryptogetcipherinfonameornid-options) |  | |
| [getCiphers](https://nodejs.org/api/crypto.html#cryptogetciphers) | ✅ | |
| [getCurves](https://nodejs.org/api/crypto.html#cryptogetcurves) | ✅ | |
| [secureHeapUsed](https://nodejs.org/api/crypto.html#cryptosecureheapused) | ✅ | |
| [setEngine](https://nodejs.org/api/crypto.html#cryptosetengineengine-flags) | ✅ | |
| [timingSafeEqual](https://nodejs.org/api/crypto.html#cryptotimingsafeequala-b) | ✅ | |

## Fips

| API | Supported? | Notes |
|-----|------------|-------|

> | [getFips](https://nodejs.org/api/crypto.html#cryptogetfips) | ✅ | |

Author: ```suggestion
| [getFips](https://nodejs.org/api/crypto.html#cryptogetfips) | ❌ |  The Workers implementation of `node:crypto` always runs in a FIPS compliant mode. This API will not be supported.  |
```

> | [fips](https://nodejs.org/api/crypto.html#cryptofips) | ✅ | Deprecated, use `getFips()` instead |

Author: ```suggestion
| [fips](https://nodejs.org/api/crypto.html#cryptofips) | ❌ |  The Workers implementation of `node:crypto` always runs in a FIPS compliant mode. This API will not be supported.  |
```

> | [setFips](https://nodejs.org/api/crypto.html#cryptosetfipsbool) | ✅ | |

Author: ```suggestion
| [setFips](https://nodejs.org/api/crypto.html#cryptosetfipsbool) | ❌ |  The Workers implementation of `node:crypto` always runs in a FIPS compliant mode. This API will not be supported.  |
```

## Random

| API | Supported? | Notes |
|-----|------------|-------|
| [getRandomValues](https://nodejs.org/api/crypto.html#cryptogetrandomvaluestypedarray) | ✅ | |
| [randomBytes](https://nodejs.org/api/crypto.html#cryptorandombytessize-callback) | ✅ | |
| [randomFillSync](https://nodejs.org/api/crypto.html#cryptorandomfillsyncbuffer-offset-size) | ✅ | |
| [randomFill](https://nodejs.org/api/crypto.html#cryptorandomfillbuffer-offset-size-callback) | ✅ | |
| [randomInt](https://nodejs.org/api/crypto.html#cryptorandomintmin-max-callback) | ✅ | |
| [randomUUID](https://nodejs.org/api/crypto.html#cryptorandomuuidoptions) | ✅ | |

## Key Derivation

| API | Supported? | Notes |
|-----|------------|-------|

> | [hkdf](https://nodejs.org/api/crypto.html#cryptohkdfdigest-ikm-salt-info-keylen-callback) | ✅ | Does not yet support KeyObject |

Author: ```suggestion
| [hkdf](https://nodejs.org/api/crypto.html#cryptohkdfdigest-ikm-salt-info-keylen-callback) | ✅ | Does not yet support an argument of KeyObject type for the `ikm` parameter. |
```

> | [hkdfSync](https://nodejs.org/api/crypto.html#cryptohkdfsyncdigest-ikm-salt-info-keylen) | ✅ | Does not yet support KeyObject |

Author: ```suggestion
| [hkdfSync](https://nodejs.org/api/crypto.html#cryptohkdfsyncdigest-ikm-salt-info-keylen) | ✅ | Does not yet support an argument of KeyObject type for the `ikm` parameter. |
```

---
pcx_content_type: configuration
title: Crypto
---

# Crypto

{{<render file="_nodejs-compat-howto.md">}}

The `node:crypto`` module provides cryptographic functionality that includes a set of wrappers for OpenSSL's hash, HMAC, cipher, decipher, sign, and verify functions.

> A subset of the the `node:crypto` module is available in Workers. All APIs in the tables below with a ✅ are supported, and unless otherwise noted, work the same way as the implementations in Node.js.

Reviewer: ```suggestion
A subset of the the `node:crypto` module is available in Workers. All APIs in the tables below with a ✅ are supported, and unless otherwise noted, work the same way as the implementations in Node.js. Additional APIs will be enabled incrementally in the future.
```

The Workers implementation of `node:crypto` always runs in a FIPS-compliant mode. As such, APIs such as `setFips()` will not be made available.

The [WebCrypto API](/workers/runtime-apis/web-crypto/) is also available within Cloudflare Workers.

> 

Author: https://github.com/cloudflare/cloudflare-docs/pull/10067/commits/9333c11e3340f14d8ec9296f3ffdf5f5438b8f6c

## Classes

| API | Supported? | Notes |
|-----|------------|-------|
| [Certificate](https://nodejs.org/api/crypto.html#class-certificate) |  | |
| [Cipher](https://nodejs.org/api/crypto.html#class-cipher) |  | |
| [Decipher](https://nodejs.org/api/crypto.html#class-decipher) |  | |

> | [DiffieHellman](https://nodejs.org/api/crypto.html#class-diffiehellman) | ✅ | |

Author: ```suggestion
| [DiffieHellman](https://nodejs.org/api/crypto.html#class-diffiehellman) | ✅ | Generating a Diffie-Hellman key (`generateKeys()`) is not supported |
```

| [createDiffieHellman(primeLength)](https://nodejs.org/api/crypto.html#cryptocreatediffiehellmanprimelength-generator) | ✅ | |
| [createDiffieHellmanGroup](https://nodejs.org/api/crypto.html#cryptocreatediffiehellmangroupname) | ✅ | |
| [createECDH](https://nodejs.org/api/crypto.html#cryptocreateecdhcurvename) |  | |
| [diffieHellman](https://nodejs.org/api/crypto.html#cryptodiffiehellmanoptions) |  | |
| [getDiffieHellman](https://nodejs.org/api/crypto.html#cryptogetdiffiehellmangroupname) | ✅ | |

## Hash

| API | Supported? | Notes |
|-----|------------|-------|
| [createHash](https://nodejs.org/api/crypto.html#cryptocreatehashalgorithm-options) | ✅ | |
| [createHmac](https://nodejs.org/api/crypto.html#cryptocreatehmacalgorithm-key-options) | ✅ | |
| [getHashes](https://nodejs.org/api/crypto.html#cryptogethashes) | ✅ | |

## Keys

| API | Supported? | Notes |
|-----|------------|-------|
| [createPrivateKey](https://nodejs.org/api/crypto.html#cryptocreateprivatekeykey) |  | |
| [createPublicKey](https://nodejs.org/api/crypto.html#cryptocreatepublickeykey) |  | |
| [createSecretKey](https://nodejs.org/api/crypto.html#cryptocreatesecretkeykey-encoding) | ✅ | |

> | [generateKey](https://nodejs.org/api/crypto.html#cryptogeneratekeytype-options-callback) | ✅ | |

Author: ```suggestion
| [generateKey](https://nodejs.org/api/crypto.html#cryptogeneratekeytype-options-callback) | ❌ | |
```

> | [generateKeyPair](https://nodejs.org/api/crypto.html#cryptogeneratekeypairtype-options-callback) | ✅ | |

Author: ```suggestion
| [generateKeyPair](https://nodejs.org/api/crypto.html#cryptogeneratekeypairtype-options-callback) | ❌ | |
```

> | [generateKeyPairSync](https://nodejs.org/api/crypto.html#cryptogeneratekeypairsynctype-options) | ✅ | |

Author: ```suggestion
| [generateKeyPairSync](https://nodejs.org/api/crypto.html#cryptogeneratekeypairsynctype-options) | ❌ | |
```

> | [generateKeySync](https://nodejs.org/api/crypto.html#cryptogeneratekeysynctype-options) | ✅ | |

Author: ```suggestion
| [generateKeySync](https://nodejs.org/api/crypto.html#cryptogeneratekeysynctype-options) | ❌ | |
```

## Sign/Verify

| API | Supported? | Notes |
|-----|------------|-------|
| [createSign](https://nodejs.org/api/crypto.html#cryptocreatesignalgorithm-options) |  | |
| [createVerify](https://nodejs.org/api/crypto.html#cryptocreateverifyalgorithm-options) |  | |
| [sign](https://nodejs.org/api/crypto.html#cryptosignalgorithm-data-key-callback) |  | |
| [verify](https://nodejs.org/api/crypto.html#cryptoverifyalgorithm-data-key-signature-callback) |  | |

## Misc

| API | Supported? | Notes |
|-----|------------|-------|
| [getCipherInfo](https://nodejs.org/api/crypto.html#cryptogetcipherinfonameornid-options) |  | |
| [getCiphers](https://nodejs.org/api/crypto.html#cryptogetciphers) | ✅ | |
| [getCurves](https://nodejs.org/api/crypto.html#cryptogetcurves) | ✅ | |

> | [secureHeapUsed](https://nodejs.org/api/crypto.html#cryptosecureheapused) | ✅ | |

Author: ```suggestion
| [secureHeapUsed](https://nodejs.org/api/crypto.html#cryptosecureheapused) | ❌ | |
```

> | [setEngine](https://nodejs.org/api/crypto.html#cryptosetengineengine-flags) | ✅ | |

Author: ```suggestion
| [setEngine](https://nodejs.org/api/crypto.html#cryptosetengineengine-flags) | ❌ | |
```

| [randomInt](https://nodejs.org/api/crypto.html#cryptorandomintmin-max-callback) | ✅ | |
| [randomUUID](https://nodejs.org/api/crypto.html#cryptorandomuuidoptions) | ✅ | |

## Key Derivation

| API | Supported? | Notes |
|-----|------------|-------|
| [hkdf](https://nodejs.org/api/crypto.html#cryptohkdfdigest-ikm-salt-info-keylen-callback) | ✅ | Does not yet support an argument of KeyObject type for the `ikm` parameter. |
| [hkdfSync](https://nodejs.org/api/crypto.html#cryptohkdfsyncdigest-ikm-salt-info-keylen) | ✅ | Does not yet support an argument of KeyObject type for the `ikm` parameter. |
| [pbkdf2](https://nodejs.org/api/crypto.html#cryptopbkdf2password-salt-iterations-keylen-digest-callback) | ✅ | |
| [pbkdf2Sync](https://nodejs.org/api/crypto.html#cryptopbkdf2password-salt-iterations-keylen-digest-callback) | ✅ | |
| [scrypt](https://nodejs.org/api/crypto.html#cryptoscryptpassword-salt-keylen-options-callback) |  | |
| [scryptSync](https://nodejs.org/api/crypto.html#cryptoscryptsyncpassword-salt-keylen-options) |  | |

## WebCrypto

| API | Supported? | Notes |
|-----|------------|-------|

> | [subtle](https://nodejs.org/api/crypto.html#cryptosubtle) | ✅ | |

Author: ```suggestion
| [subtle](/workers/runtime-apis/web-crypto/#subtlecrypto-methods) | ✅ | |
```

> | [webcrypto](https://nodejs.org/api/crypto.html#) | ✅ | |

Author: ```suggestion
| [webcrypto](/workers/runtime-apis/web-crypto/) | ✅ | |
```
